In [30]:
import pandas as pd
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

# !conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Solving environment: done

# All requested packages already installed.

Libraries imported.


## Scraping Data from Wikipedia

In [31]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]
df

Postal code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
7           M8A      Not assigned   
8           M9A         Etobicoke   
9           M1B       Scarborough   
10          M2B      Not assigned   
11          M3B        North York   
12          M4B         East York   
13          M5B  Downtown Toronto   
14          M6B        North York   
15          M7B      Not assigned   
16          M8B      Not assigned   
17          M9B         Etobicoke   
18          M1C       Scarborough   
19          M2C      Not assigned   
20          M3C        North York   
21          M4C         East York   
22          M5C  Downtown Toronto   
23          M6C              York   
24          M7C      Not assigned   
25          M8C      Not assigned   
26          M9C         Etobicoke   
27          M1E       Scarborough   
28          M2E      Not assigned   
29          M3E      Not assigned   
30          M4E      East Toronto   
31          M5E  Downtown Toronto   
32          M6E              York   
33          M7E      Not assigned   
34          M8E      Not assigned   
35          M9E      Not assigned   
36          M1G       Scarborough   
37          M2G      Not assigned   
38          M3G      Not assigned   
39          M4G         East York   
40          M5G  Downtown Toronto   
41          M6G  Downtown Toronto   
42          M7G      Not assigned   
43          M8G      Not assigned   
44          M9G      Not assigned   
45          M1H       Scarborough   
46          M2H        North York   
47          M3H        North York   
48          M4H         East York   
49          M5H  Downtown Toronto   
50          M6H      West Toronto   
51          M7H      Not assigned   
52          M8H      Not assigned   
53          M9H      Not assigned   
54          M1J       Scarborough   
55          M2J        North York   
56          M3J        North York   
57          M4J         East York   
58          M5J  Downtown Toronto   
59          M6J      West Toronto   
60          M7J      Not assigned   
61          M8J      Not assigned   
62          M9J      Not assigned   
63          M1K       Scarborough   
64          M2K        North York   
65          M3K        North York   
66          M4K      East Toronto   
67          M5K  Downtown Toronto   
68          M6K      West Toronto   
69          M7K      Not assigned   
70          M8K      Not assigned   
71          M9K      Not assigned   
72          M1L       Scarborough   
73          M2L        North York   
74          M3L        North York   
75          M4L      East Toronto   
76          M5L  Downtown Toronto   
77          M6L        North York   
78          M7L      Not assigned   
79          M8L      Not assigned   
80          M9L        North York   
81          M1M       Scarborough   
82          M2M        North York   
83          M3M        North York   
84          M4M      East Toronto   
85          M5M        North York   
86          M6M              York   
87          M7M      Not assigned   
88          M8M      Not assigned   
89          M9M        North York   
90          M1N       Scarborough   
91          M2N        North York   
92          M3N        North York   
93          M4N   Central Toronto   
94          M5N   Central Toronto   
95          M6N              York   
96          M7N      Not assigned   
97          M8N      Not assigned   
98          M9N              York   
99          M1P       Scarborough   
100         M2P        North York   
101         M3P      Not assigned   
102         M4P   Central Toronto   
103         M5P   Central Toronto   
104         M6P      West Toronto   
105         M7P      Not assigned   
106         M8P      Not assigned   
107     

## Dropping NaN values and Renaming column to PostalCode

In [32]:
df.dropna(axis=0,inplace=True)
df.rename(columns={'Postal code':'PostalCode'}, inplace=True)

### Separating two or more neighborhoods with comma

In [33]:
df["Neighborhood"] = df["Neighborhood"].str.replace(" /",",")


In [34]:
df

PostalCode           Borough  \
2          M3A        North York   
3          M4A        North York   
4          M5A  Downtown Toronto   
5          M6A        North York   
6          M7A  Downtown Toronto   
8          M9A         Etobicoke   
9          M1B       Scarborough   
11         M3B        North York   
12         M4B         East York   
13         M5B  Downtown Toronto   
14         M6B        North York   
17         M9B         Etobicoke   
18         M1C       Scarborough   
20         M3C        North York   
21         M4C         East York   
22         M5C  Downtown Toronto   
23         M6C              York   
26         M9C         Etobicoke   
27         M1E       Scarborough   
30         M4E      East Toronto   
31         M5E  Downtown Toronto   
32         M6E              York   
36         M1G       Scarborough   
39         M4G         East York   
40         M5G  Downtown Toronto   
41         M6G  Downtown Toronto   
45         M1H       Scarborough   
46         M2H        North York   
47         M3H        North York   
48         M4H         East York   
49         M5H  Downtown Toronto   
50         M6H      West Toronto   
54         M1J       Scarborough   
55         M2J        North York   
56         M3J        North York   
57         M4J         East York   
58         M5J  Downtown Toronto   
59         M6J      West Toronto   
63         M1K       Scarborough   
64         M2K        North York   
65         M3K        North York   
66         M4K      East Toronto   
67         M5K  Downtown Toronto   
68         M6K      West Toronto   
72         M1L       Scarborough   
73         M2L        North York   
74         M3L        North York   
75         M4L      East Toronto   
76         M5L  Downtown Toronto   
77         M6L        North York   
80         M9L        North York   
81         M1M       Scarborough   
82         M2M        North York   
83         M3M        North York   
84         M4M      East Toronto   
85         M5M        North York   
86         M6M              York   
89         M9M        North York   
90         M1N       Scarborough   
91         M2N        North York   
92         M3N        North York   
93         M4N   Central Toronto   
94         M5N   Central Toronto   
95         M6N              York   
98         M9N              York   
99         M1P       Scarborough   
100        M2P        North York   
102        M4P   Central Toronto   
103        M5P   Central Toronto   
104        M6P      West Toronto   
107        M9P         Etobicoke   
108        M1R       Scarborough   
109        M2R        North York   
111        M4R   Central Toronto   
112        M5R   Central Toronto   
113        M6R      West Toronto   
114        M7R       Mississauga   
116        M9R         Etobicoke   
117        M1S       Scarborough   
120        M4S   Central Toronto   
121        M5S  Downtown Toronto   
122        M6S      West Toronto   
126        M1T       Scarborough   
129        M4T   Central Toronto   
130        M5T  Downtown Toronto   
135        M1V       Scarborough   
138        M4V   Central Toronto   
139        M5V  Downtown Toronto   
142        M8V         Etobicoke   
143        M9V         Etobicoke   
144        M1W       Scarborough   
147        M4W  Downtown Toronto   
148        M5W  Downtown Toronto   
151        M8W         Etobicoke   
152        M9W         Etobicoke   
153        M1X       Scarborough   
156        M4X  Downtown Toronto   
157        M5X  Downtown Toronto   
160        M8X         Etobicoke   
165        M4Y  Downtown Toronto   
168        M7Y      East Toronto   
169        M8Y         Etobicoke   
178        M8Z         Etobicoke   

                                          Neighborhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence 

In [35]:
df.shape

(103, 3)

## Importing coordinates to dataframe and renaming colum to PostalCode

In [36]:
coordinates = pd.read_csv('Geospatial_Coordinates.csv')
coordinates.rename(columns={'Postal Code':'PostalCode'}, inplace=True)

## Merging coordinates with toronto data

In [37]:
neighborhoods = pd.merge(df, coordinates, how='inner', on='PostalCode')

Quickly examine the resulting dataframe.

In [38]:
neighborhoods.head()

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

And make sure that the dataset has all 10 boroughs and 103 neighborhoods.


In [39]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(neighborhoods['Borough'].unique()),
        neighborhoods.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of New York City.

In order to define an instance of the geocoder, we need to define a user_agent. We will name our agent <em>ny_explorer</em>, as shown below.

In [40]:
address = 'Toronto'

geolocator = Nominatim(user_agent="toronto_explorer", timeout=None)
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.

In [42]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(neighborhoods['Latitude'], neighborhoods['Longitude'], neighborhoods['Borough'], neighborhoods['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

Folium is a great visualization library. Feel free to zoom into the above map, and click on each circle mark to reveal the name of the neighborhood and its respective borough.

However, for illustration purposes, let's simplify the above map and segment and cluster only the neighborhoods in Downtown Toronto. So let's slice the original dataframe and create a new dataframe of the Downtown Toronto data.

In [43]:
downtown_data = neighborhoods[neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.644771 -79.373306

Let's get the geographical coordinates of Downtown Toronto.

In [44]:
address = 'Downtown Toronto, Ontario'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6563221, -79.3809161.


As we did with all of New York City, let's visualizat Downtown Toronto the neighborhoods in it.

In [46]:
# create map of Downtown Toronto using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighborhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

In [77]:
CLIENT_ID = 'your-client-ID' # your Foursquare ID
CLIENT_SECRET = 'your-client-secret' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: your-client-ID
CLIENT_SECRET:your-client-secret


Let's explore the first neighborhood in our dataframe.
Get the neighborhood's name.

In [48]:
downtown_data.loc[0, 'Neighborhood']

'Regent Park, Harbourfront'

In [49]:
neighborhood_latitude = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = downtown_data.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Regent Park, Harbourfront are 43.6542599, -79.3606359.


First, let's create the GET request URL. Name your URL **url**.

In [51]:
# type your answer here
radius = 500 
LIMIT = 100
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

Get results

In [52]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5e9ee59198205d001b7944a3'},
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Corktown',
  'headerFullLocation': 'Corktown, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 46,
  'suggestedBounds': {'ne': {'lat': 43.6587599045, 'lng': -79.3544279001486},
   'sw': {'lat': 43.6497598955, 'lng': -79.36684389985142}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '54ea41ad498e9a11e9e13308',
       'name': 'Roselle Desserts',
       'location': {'address': '362 King St E',
        'crossStreet': 'Trinity St',
        'lat': 43.653446723052674,
        'lng': -79.3620167174383,
        'labeledLatLngs': [{'label': 'display',
 

From the Foursquare lab in the previous module, we know that all the information is in the items key. Before we proceed, let's borrow the get_category_type function from the Foursquare lab

In [53]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [54]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

/Users/nomi/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: FutureWarning: pandas.io.json.json_normalize is deprecated, use pandas.json_normalize instead
  This is separate from the ipykernel package so we can avoid doing imports until


name           categories        lat        lng
0        Roselle Desserts               Bakery  43.653447 -79.362017
1           Tandem Coffee          Coffee Shop  43.653559 -79.361809
2  Cooper Koo Family YMCA  Distribution Center  43.653249 -79.358008
3     Body Blitz Spa East                  Spa  43.654735 -79.359874
4      Morning Glory Cafe       Breakfast Spot  43.653947 -79.361149

And how many venues were returned by Foursquare?

In [55]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

46 venues were returned by Foursquare.


## 2. Explore Neighborhoods in Manhattan

#### Let's create a function to repeat the same process to all the neighborhoods in Manhattan


In [56]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

#### Now write the code to run the above function on each neighborhood and create a new dataframe called *downtown_venues*.

In [57]:
# type your answer here

downtown_venues = getNearbyVenues(names=downtown_data['Neighborhood'],
                                   latitudes=downtown_data['Latitude'],
                                   longitudes=downtown_data['Longitude']
                                  )



Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town, Cabbagetown
First Canadian Place, Underground city
Church and Wellesley


#### Let's check the size of the resulting dataframe

In [58]:
print(downtown_venues.shape)
downtown_venues.head()

(1228, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4      Morning Glory Cafe       43.653947       -79.361149   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4       Breakfast Spot

Let's check how many venues were returned for each neighborhood

In [59]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
CN Tower, King and Spadina, Railway Lands, Harb...                     16   
Central Bay Street                                                     65   
Christie                                                               18   
Church and Wellesley                                                   72   
Commerce Court, Victoria Hotel                                        100   
First Canadian Place, Underground city                                100   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
Kensington Market, Chinatown, Grange Park                              56   
Queen's Park, Ontario Provincial Government                            38   
Regent Park, Harbourfront                                              46   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
St. James Town                                                         84   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         97   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         35   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
CN Tower, King and Spadina, Railway Lands, Harb...                      16   
Central Bay Street                                                      65   
Christie                                                                18   
Church and Wellesley                                                    72   
Commerce Court, Victoria Hotel                                         100   
First Canadian Place, Underground city                                 100   
Garden District, Ryerson                                               100   
Harbourfront East, Union Station, Toronto Islands                      100   
Kensington Market, Chinatown, Grange Park                               56   
Queen's Park, Ontario Provincial Government                             38   
Regent Park, Harbourfront                                               46   
Richmond, Adelaide, King                                                94   
Rosedale                                                                 4   
St. James Town                                                          84   
St. James Town, Cabbagetown                                             46   
Stn A PO Boxes                                                          97   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                          35   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Berczy Park                                            57              57   
CN Tower, King and Spadina, Railway Lands, Harb...     16              16   
Central Bay Street                                     65              65   
Christie                                               18              18   
Church and Wellesley                                   72              72   
Commerce Court, Victoria Hotel                        100             100   
First Canadian Place, Underground city                100             100   
Garden District, Ryerson                              100             100   
Harbourfront East, Union 

#### Let's find out how many unique categories can be curated from all the returned venues

In [60]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 206 uniques categories.


## 3. Analyze Each Neighborhood


In [61]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                   0          0           0           0       1     0    0   
1                   0          0           0           0       0     0    0   
2                   0          0           0           0       0     0    0   
3                   0          0           0           0       0     0    0   
4                   0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0           0                   0       0              0          0         0   
1           0                   0       0              0          0         0   
2           0                   0       0              0          0         0   
3           0                   0       0              0          0         0   
4           0                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               1        0                0         0   

   Burger Joint  Burrito Place  Business Service  Butcher  Café  Candy Store  \
0             0              0                 0        0     0            0   
1             0              0                 0        0     0            0   
2             0              0                 0        0     0            0   
3             0              0                 0        0     0            0   
4             0              0                 0        0     0            0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2    

And let's examine the new dataframe size.

In [62]:
downtown_onehot.shape

(1228, 206)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [63]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped

Neighborhood  Yoga Studio  Airport  \
0                                         Berczy Park     0.000000   0.0000   
1   CN Tower, King and Spadina, Railway Lands, Har...     0.000000   0.0625   
2                                  Central Bay Street     0.015385   0.0000   
3                                            Christie     0.000000   0.0000   
4                                Church and Wellesley     0.027778   0.0000   
5                      Commerce Court, Victoria Hotel     0.000000   0.0000   
6              First Canadian Place, Underground city     0.000000   0.0000   
7                            Garden District, Ryerson     0.000000   0.0000   
8   Harbourfront East, Union Station, Toronto Islands     0.000000   0.0000   
9           Kensington Market, Chinatown, Grange Park     0.000000   0.0000   
10        Queen's Park, Ontario Provincial Government     0.026316   0.0000   
11                          Regent Park, Harbourfront     0.021739   0.0000   
12                           Richmond, Adelaide, King     0.000000   0.0000   
13                                           Rosedale     0.000000   0.0000   
14                                     St. James Town     0.000000   0.0000   
15                        St. James Town, Cabbagetown     0.000000   0.0000   
16                                     Stn A PO Boxes     0.010309   0.0000   
17           Toronto Dominion Centre, Design Exchange     0.000000   0.0000   
18                     University of Toronto, Harbord     0.028571   0.0000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0               0.0000        0.0000           0.000            0.000   
1               0.0625        0.0625           0.125            0.125   
2               0.0000        0.0000           0.000            0.000   
3               0.0000        0.0000           0.000            0.000   
4               0.0000        0.0000           0.000            0.000   
5               0.0000        0.0000           0.000            0.000   
6               0.0000        0.0000           0.000            0.000   
7               0.0000        0.0000           0.000            0.000   
8               0.0000        0.0000           0.000            0.000   
9               0.0000        0.0000           0.000            0.000   
10              0.0000        0.0000           0.000            0.000   
11              0.0000        0.0000           0.000            0.000   
12              0.0000        0.0000           0.000            0.000   
13              0.0000        0.0000           0.000            0.000   
14              0.0000        0.0000           0.000            0.000   
15              0.0000        0.0000           0.000            0.000   
16              0.0000        0.0000           0.000            0.000   
17              0.0000        0.0000           0.000            0.000   
18              0.0000        0.0000           0.000            0.000   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0              0.000             0.000000      0.000000      0.00   
1              0.125             0.000000      0.000000      0.00   
2              0.000             0.000000      0.000000      0.00   
3              0.000             0.000000      0.000000      0.00   
4              0.000             0.013889      0.000000      0.00   
5              0.000             0.040000      0.000000      0.00   
6              0.000             0.030000      0.000000      0.00   
7              0.000             0.010000      0.000000      0.00   
8              0.000             0.000000      0.000000      0.05   
9              0.000             0.000000      0.000000      0.00   
10             0.000             0.000000      0.000000      0.00   
11             0.000             0.000000      0.021739      0.00   
12             0.000             0.031915      0.000000      0.00   
13             0.000             

#### Let's confirm the new size


In [64]:
downtown_grouped.shape

(19, 206)

#### Let's print each neighborhood along with the top 5 most common venues

In [65]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2  Italian Restaurant  0.04
3  Seafood Restaurant  0.04
4                Café  0.04


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0    Airport Lounge  0.12
1   Airport Service  0.12
2  Airport Terminal  0.12
3     Boat or Ferry  0.06
4           Airport  0.06


----Central Bay Street----
                 venue  freq
0          Coffee Shop  0.20
1                 Café  0.06
2   Italian Restaurant  0.06
3  Japanese Restaurant  0.05
4       Sandwich Place  0.05


----Christie----
                venue  freq
0       Grocery Store  0.22
1                Café  0.17
2                Park  0.11
3  Italian Restaurant  0.06
4               Diner  0.06


----Church and Wellesley----
                 venue  freq
0          Coffee Shop  0.07
1     Sushi Restaurant  0.07
2  Japanese Restaurant  0.06
3 

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [66]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [68]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
2                                 Central Bay Street           Coffee Shop   
3                                           Christie         Grocery Store   
4                               Church and Wellesley      Sushi Restaurant   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar    Seafood Restaurant    Italian Restaurant   
1       Airport Service      Airport Terminal      Sculpture Garden   
2                  Café    Italian Restaurant   Japanese Restaurant   
3                  Café                  Park                 Diner   
4           Coffee Shop   Japanese Restaurant            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0              Beer Bar                Bakery        Farmers Market   
1       Harbor / Marina                 Plane              Boutique   
2        Sandwich Place        Ice Cream Shop       Bubble Tea Shop   
3    Athletics & Sports           Candy Store    Italian Restaurant   
4             Gastropub                   Pub            Smoke Shop   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop                       Café             Restaurant  
1   Rental Car Location                Coffee Shop          Boat or Ferry  
2          Burger Joint  Middle Eastern Restaurant            Salad Place  
3            Restaurant                 Baby Store            Gas Station  
4                 Hotel                Yoga Studio           Burger Joint

## 4. Cluster Neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [69]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([1, 3, 4, 0, 1, 1, 1, 1, 1, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [70]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

downtown_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M5E  Downtown Toronto                                  Berczy Park   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               1           Coffee Shop   
1  43.662301 -79.389494               4           Coffee Shop   
2  43.657162 -79.378937               1        Clothing Store   
3  43.651494 -79.375418               1           Coffee Shop   
4  43.644771 -79.373306               1           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery                   Pub                  Park   
1      Sushi Restaurant                 Diner           Yoga Studio   
2           Coffee Shop                  Café   Japanese Restaurant   
3                  Café          Cocktail Bar             Gastropub   
4          Cocktail Bar    Seafood Restaurant    Italian Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant        Breakfast Spot                  Café   
1          Burger Joint                   Bar              Beer Bar   
2        Cosmetics Shop       Bubble Tea Shop            Restaurant   
3    Italian Restaurant   American Restaurant                 Hotel   
4              Beer Bar                Bakery        Farmers Market   

       8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Theater           Yoga Studio         Cosmetics Shop  
1         Italian Restaurant             Juice Bar    Distribution Center  
2  Middle Eastern Restaurant             Bookstore       Ramen Restaurant  
3                   Creperie    Seafood Restaurant         Lingerie Store  
4                Cheese Shop                  Café             Restaurant

Finally, let's visualize the resulting clusters

In [71]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighborhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.

#### Cluster 1


In [72]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
6   Downtown Toronto               0         Grocery Store   
11  Downtown Toronto               0                  Café   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
6                   Café                  Park                 Diner   
11            Restaurant                   Bar    Italian Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
6     Athletics & Sports           Candy Store    Italian Restaurant   
11   Japanese Restaurant             Bookstore                Bakery   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
6             Restaurant            Baby Store            Gas Station  
11           Yoga Studio              Beer Bar             Beer Store

#### Cluster 2


In [73]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               1           Coffee Shop   
2   Downtown Toronto               1        Clothing Store   
3   Downtown Toronto               1           Coffee Shop   
4   Downtown Toronto               1           Coffee Shop   
7   Downtown Toronto               1           Coffee Shop   
8   Downtown Toronto               1           Coffee Shop   
9   Downtown Toronto               1           Coffee Shop   
10  Downtown Toronto               1           Coffee Shop   
12  Downtown Toronto               1                  Café   
15  Downtown Toronto               1           Coffee Shop   
16  Downtown Toronto               1           Coffee Shop   
17  Downtown Toronto               1           Coffee Shop   
18  Downtown Toronto               1      Sushi Restaurant   

   2nd Most Common Venue  3rd Most Common Venue 4th Most Common Venue  \
0                 Bakery                    Pub                  Park   
2            Coffee Shop                   Café   Japanese Restaurant   
3                   Café           Cocktail Bar             Gastropub   
4           Cocktail Bar     Seafood Restaurant    Italian Restaurant   
7                   Café             Restaurant        Clothing Store   
8               Aquarium             Restaurant    Italian Restaurant   
9                  Hotel                   Café            Restaurant   
10                  Café             Restaurant                 Hotel   
12           Coffee Shop  Vietnamese Restaurant    Mexican Restaurant   
15    Italian Restaurant     Seafood Restaurant                  Café   
16            Restaurant                   Park                  Café   
17                  Café                  Hotel            Restaurant   
18           Coffee Shop    Japanese Restaurant            Restaurant   

   5th Most Common Venue 6th Most Common Venue          7th Most Common Venue  \
0             Restaurant        Breakfast Spot                           Café   
2         Cosmetics Shop       Bubble Tea Shop                     Restaurant   
3     Italian Restaurant   American Restaurant                          Hotel   
4               Beer Bar                Bakery                 Farmers Market   
7                  Hotel   American Restaurant                            Gym   
8                  Hotel                  Café            Fried Chicken Joint   
9    American Restaurant   Japanese Restaurant             Seafood Restaurant   
10                   Gym   American Restaurant                  Deli / Bodega   
12         Grocery Store           Gaming Cafe  Vegetarian / Vegan Restaurant   
15                 Hotel          Cocktail Bar            Japanese Restaurant   
16                   Pub    Italian Restaurant                    Pizza Place   
17            Steakhouse   American Restaurant                            Bar   
18             Gastropub                   Pub                     Smoke Shop   

        8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                     Theater           Yoga Studio         Cosmetics Shop  
2   Middle Eastern Restaurant             Bookstore       Ramen Restaurant  
3                    Creperie    Seafood Restaurant         Lingerie Store  
4                 Cheese Shop                  Café             Restaurant  
7             Thai Restaurant         Deli / Bodega             Steakhouse  
8              Scenic Lookout   Sporting Goods Shop                Brewery  
9                 Salad Place         Deli / Bodega     Italian Restaurant  
10         Italian Restaurant   Japanese Restaurant     Seafood Restaurant  
12               Dessert Shop                   Bar                 Bakery  
15                   Beer Bar            Restaurant         Breakfast Spot  
16                     Bakery  Caribbean Restaurant         Breakfast Spot  
17         Seafood Restaurant      Asian Restaurant          D

#### Cluster 3


In [74]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14                 Trail            Playground         Women's Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14              Creperie      Doner Restaurant               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14   Distribution Center        Discount Store                  Diner

#### Cluster 4


In [75]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
13  Downtown Toronto               3        Airport Lounge   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
13       Airport Service      Airport Terminal      Sculpture Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
13       Harbor / Marina                 Plane              Boutique   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
13   Rental Car Location           Coffee Shop          Boat or Ferry

#### Cluster 5


In [76]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1  Downtown Toronto               4           Coffee Shop   
5  Downtown Toronto               4           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1      Sushi Restaurant                 Diner           Yoga Studio   
5                  Café    Italian Restaurant   Japanese Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1          Burger Joint                   Bar              Beer Bar   
5        Sandwich Place        Ice Cream Shop       Bubble Tea Shop   

  8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
1    Italian Restaurant                  Juice Bar    Distribution Center  
5          Burger Joint  Middle Eastern Restaurant            Salad Place